### PR TITLE
refactor: introduce `BuildStage`

### PR DESCRIPTION
**Pull request by hyf0**
### Description

Part of #237.

Rolldown will have three stages in bundling:
1. building
2. linking
3. bundling/generating

### Test Plan

---

### CODE REVIEW

**Comment by hyf0** [comment]
Current dependencies on/for this PR:
* main
  * **PR #244** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/244?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #245** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/245?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/rolldown-rs/rolldown/244?utm_source=stack-comment).

File: crates/rolldown/src/bundler/stages/build_stage.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    self.plugin_driver.build_start().await?;
+
+    let build_info = self.build_inner(fs).await;
+
+    if let Err(e) = build_info {
+      let error = e.get().expect("should have a error");
+      self
+        .plugin_driver
+        .build_end(Some(&HookBuildEndArgs {
+          // TODO(hyf0): 1.Need a better way to expose the error, 2.How to handle multiple errors
```
Comment (hyf0) [inline crates/rolldown/src/bundler/stages/build_stage.rs:102]: Rollup only emit the first error from @underfin

File: crates/rolldown/src/bundler/stages/build_stage.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+          error: format!("{:?}\n{:?}", error.code(), error.to_diagnostic().print_to_string()),
+        }))
+        .await?;
+      return Err(e);
+    }
+
+    self.plugin_driver.build_end(None).await?;
```
Comment (hyf0) [inline crates/rolldown/src/bundler/stages/build_stage.rs:109]: The `build_end` hook should placing in the end of `linking` from @underfin

File: crates/rolldown/src/bundler/stages/build_stage.rs (hunk 1 above)
Comment (underfin) [reply crates/rolldown/src/bundler/stages/build_stage.rs:102]: Rollup here using early error, if meet error will exit.The argument is `Error`.

**Review by underfin — APPROVED** [review]

**Comment by hyf0** [comment]
### Merge activity

* **Nov 13, 5:41 AM**: @hyf0 started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/rolldown-rs/rolldown/244).
* **Nov 13, 5:41 AM**: @hyf0 merged this pull request with [Graphite](https://app.graphite.dev/github/pr/rolldown-rs/rolldown/244).

**Pull request merged**